### PR TITLE
LSNBLDR-633: Restrict editing Lessons pages/subpages to one person

### PR DESCRIPTION
**Pull request by nicholaswilson100**

### CODE REVIEW

**Comment by nicholaswilson100** [comment]
There are comments on code relating to this PR on a couple of other PRs: https://github.com/sakaiproject/sakai/pull/3581 and https://github.com/sakaiproject/sakai/pull/3384.

File: lessonbuilder/hbm/src/java/org/sakaiproject/lessonbuildertool/SimplePageImpl.java [hunk 1]
```diff
@@ -179,4 +181,13 @@ public void setFolder(String f) {
 		this.folder = f;
 	}
 
+	public boolean isOwned() {
+		return owned;
+	}
+
+	public void setOwned(Boolean owned) {
+		if(owned == null)
+			owned = false;
```
Comment (bjones86) [inline lessonbuilder/hbm/src/java/org/sakaiproject/lessonbuildertool/SimplePageImpl.java]: Please use braces on single line conditionals.

File: lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/beans/SimplePageBean.java [hunk 2]
```diff
@@ -28,6 +28,7 @@
 import java.text.Format;
 import java.math.BigDecimal;
 import java.util.Locale;
+import org.sakaiproject.user.api.*;
```
Comment (bjones86) [inline lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/beans/SimplePageBean.java]: Please do not use * imports

File: lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/beans/SimplePageBean.java [hunk 3]
```diff
@@ -3676,7 +3690,7 @@ public SecurityAdvice isAllowed(String userId, String function, String reference
     // the group list, so you need to do i.setGroups().
        public List<String> setItemGroups (SimplePageItem i, String[] groups) {
 	   // can't allow groups on student pages
-	   if (getCurrentPage().getOwner() != null)
+	   if (isStudentPage(getCurrentPage()))
 	       return null;
```
Comment (bjones86) [inline lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/beans/SimplePageBean.java]: While you're on this line, you could improve this single line conditional to include braces

File: lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/beans/SimplePageBean.java [hunk 4]
```diff
@@ -5966,7 +6005,7 @@ public Path getPagePath(SimplePage page, Set<Long>seen) {
 		    SimplePage p = simplePageToolDao.getPage(i.getPageId());
 		    if (p == null)
 			continue;
-		    if (p.getOwner() != null)  // probably can't happen
+		    if (isStudentPage(p))  // probably can't happen
```
Comment (bjones86) [inline lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/beans/SimplePageBean.java]: Same here

File: lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/beans/SimplePageBean.java [hunk 5]
```diff
@@ -6723,7 +6762,7 @@ public String addComment() {
 		}
 
 		// student page
-		boolean isStudent = (currentPage.getOwner() != null);
+		boolean isStudent = (isStudentPage(currentPage));
```
Comment (bjones86) [inline lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/beans/SimplePageBean.java]: Outer parenthesis not necessary here.

File: lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/beans/SimplePageBean.java [hunk 6]
```diff
@@ -6889,7 +6928,7 @@ public String deleteComment(String commentUUID) {
 	}
 	
 	public void addStudentContentSection() {
-		if (!canEditPage()) 
+		if (!canEditPage())
```
Comment (bjones86) [inline lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/beans/SimplePageBean.java]: Braces...

File: lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/producers/ChecklistProducer.java [hunk 7]
```diff
@@ -149,7 +149,7 @@ public void fillComponents(UIContainer tofill, ViewParameters viewparams, Compon
 	    UIInput customCssClass = UIInput.make(form, "customCssClass", "#{simplePageBean.customCssClass}", customClass);
 	    UIOutput.make(form, "custom-css-label", messageLocator.getMessage("simplepage.custom.css.class"));
 
-            if (page.getOwner() == null)
+            if (!simplePageBean.isStudentPage(page))
```
Comment (bjones86) [inline lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/producers/ChecklistProducer.java]: Braces...

File: lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/producers/EditPageProducer.java [hunk 8]
```diff
@@ -155,8 +155,8 @@ public void fillComponents(UIContainer tofill, ViewParameters viewparams, Compon
 
 			richTextEvolver.evolveTextInput(instructions);
 
-			if (page.getOwner() == null) {
-			    // these options don't apply on student pages
+			if (!simplePageBean.isStudentPage(page)) {
+				// these options don't apply on student pages
```
Comment (bjones86) [inline lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/producers/EditPageProducer.java]: Whitespace is inconsistent here

File: lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/producers/ShowPageProducer.java [hunk 9]
```diff
@@ -39,21 +39,7 @@
...
-import java.util.HashMap;
-import java.util.List;
-import java.util.Set;
-import java.util.HashSet;
-import java.util.Locale;
-import java.util.Map;
-import java.util.StringTokenizer;
-import java.util.TimeZone;
-import java.util.Calendar;
+import java.util.*;
```
Comment (bjones86) [inline lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/producers/ShowPageProducer.java]: Please set your IDE to not alter imports; * imports are not terribly good practise

File: lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/producers/ShowPageProducer.java [hunk 10]
```diff
@@ -66,6 +52,7 @@
 import javax.crypto.spec.SecretKeySpec;
 import javax.xml.bind.DatatypeConverter;
 
+import org.apache.commons.lang.*;
```
Comment (bjones86) [inline lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/producers/ShowPageProducer.java]: Please no * imports

File: lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/producers/ShowPageProducer.java [hunk 11]
```diff
@@ -3183,7 +3176,7 @@ else if(questionStatus == Status.FAILED)
 				// default placeholder
 				// we know the defaults exist because we include them, so
 				// we never need to consider default general
-				if (currentPage.getOwner() != null)
+				if (simplePageBean.isStudentPage(currentPage))
```
Comment (bjones86) [inline lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/producers/ShowPageProducer.java]: Braces

File: lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/producers/ShowPageProducer.java [hunk 12]
```diff
@@ -3924,7 +3942,7 @@ private void createEditItemDialog(UIContainer tofill, SimplePage currentPage, Si
 		UIInput.make(form, "description", "#{simplePageBean.description}");
 
 		UIOutput changeDiv = UIOutput.make(form, "changeDiv");
-		if(currentPage.getOwner() != null) changeDiv.decorate(new UIStyleDecorator("noDisplay"));
+		if(simplePageBean.isStudentPage(currentPage)) changeDiv.decorate(new UIStyleDecorator("noDisplay"));
```
Comment (bjones86) [inline lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/producers/ShowPageProducer.java]: Braces

File: lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/producers/ShowPageProducer.java [hunk 13]
```diff
@@ -3984,7 +4002,7 @@ private void createEditItemDialog(UIContainer tofill, SimplePage currentPage, Si
 		UIInput.make(form, "item-id", "#{simplePageBean.itemId}");
 
 		UIOutput permDiv = UIOutput.make(form, "permDiv");
-		if(currentPage.getOwner() != null) permDiv.decorate(new UIStyleDecorator("noDisplay"));
+		if(simplePageBean.isStudentPage(currentPage)) permDiv.decorate(new UIStyleDecorator("noDisplay"));
```
Comment (bjones86) [inline lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/producers/ShowPageProducer.java]: Braces

File: lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/producers/ShowPageProducer.java [hunk 14]
```diff
@@ -4029,7 +4047,7 @@ private void createEditItemDialog(UIContainer tofill, SimplePage currentPage, Si
 
 		// can't use site groups on user content, and don't want students to hack
 		// on groups for site content
-		if (currentPage.getOwner() == null)
+		if (!simplePageBean.isStudentPage(currentPage))
```
Comment (bjones86) [inline lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/producers/ShowPageProducer.java]: Braces

File: lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/producers/ShowPageProducer.java [hunk 15]
```diff
@@ -4430,7 +4448,7 @@ private void createEditTitleDialog(UIContainer tofill, SimplePage page, SimplePa
 		}
 
 		UIOutput gradeBook = UIOutput.make(form, "gradeBookDiv");
-		if(page.getOwner() != null) gradeBook.decorate(new UIStyleDecorator("noDisplay"));
+		if(simplePageBean.isStudentPage(page)) gradeBook.decorate(new UIStyleDecorator("noDisplay"));
```
Comment (bjones86) [inline lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/producers/ShowPageProducer.java]: Braces

File: lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/producers/ShowPageProducer.java [hunk 16]
```diff
@@ -4470,11 +4488,51 @@ private void createEditTitleDialog(UIContainer tofill, SimplePage page, SimplePa
...
 			UIOutput.make(form, "cssUpload");
 			UIBoundBoolean.make(form, "nodownloads", 
 					    "#{simplePageBean.nodownloads}", 
 					    (simplePageBean.getCurrentSite().getProperties().getProperty("lessonbuilder-nodownloadlinks") != null));
+
+			//Set the changeOwner dropdown in the settings dialog
+			UIOutput.make(form, "changeOwnerSection");
+			List<String> roleOptions = new ArrayList<String>();
+			List<String> roleLabels = new ArrayList<String>();
+				List<String> possOwners = new LinkedList<String>();
```
Comment (bjones86) [inline lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/producers/ShowPageProducer.java]: Whitespace is off in the below section

File: lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/producers/ShowPageProducer.java [hunk 17]
```diff
@@ -4470,11 +4488,51 @@ private void createEditTitleDialog(UIContainer tofill, SimplePage page, SimplePa
...
+					.replace("{1}", messageLocator.getMessage("simplepage.permissions")).replace("{2}", messageLocator.getMessage("simplepage.more-tools")));
 			UIOutput.make(form, "cssUploadLabel", messageLocator.getMessage("simplepage.css-upload-label"));
 			UIOutput.make(form, "cssUpload");
 			UIBoundBoolean.make(form, "nodownloads", 
 					    "#{simplePageBean.nodownloads}", 
 					    (simplePageBean.getCurrentSite().getProperties().getProperty("lessonbuilder-nodownloadlinks") != null));
+
+			//Set the changeOwner dropdown in the settings dialog
+			UIOutput.make(form, "changeOwnerSection");
+			List<String> roleOptions = new ArrayList<String>();
```
Comment (bjones86) [inline lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/producers/ShowPageProducer.java]: Diamond operators here and below

File: lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/producers/ShowPageProducer.java [hunk 18]
```diff
@@ -4470,11 +4488,51 @@ private void createEditTitleDialog(UIContainer tofill, SimplePage page, SimplePa
...
+						roleOptions.add(owner);
+						roleLabels.add(getUserDisplayName(owner));
+					}
+				}
+				else {
+					roleOptions.add(null);
+					roleLabels.add(messageLocator.getMessage("simplepage.default-user"));
+				}
+
+				for(Iterator<String> users = possOwners.iterator(); users.hasNext();){
```
Comment (bjones86) [inline lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/producers/ShowPageProducer.java]: `Iterator` not necessary here

File: lessonbuilder/tool/src/webapp/templates/ShowPage.html [hunk 19]
```diff
@@ -117,7 +117,7 @@ <h2 ><span rsf:id="crumb:"><a style="background:none"  href="#" rsf:id="crumb-li
 				 <img rsf:id="resetimage" src="/library/image/transparent.gif"/>
 			     </a>
 				   <h2>
-				       <span rsf:id="pagetitle"></span><a role="button" class="portlet title-tools title reset" rsf:id="pagetitlelink"><span class="Mrphs-toolTitleNav__link Mrphs-toolTitleNav__link--reset"></span><span rsf:id="pagetitletext"></span></a>
+					 <div id="owner" rsf:id="owner"></div><span rsf:id="pagetitle"></span><a role="button" class="portlet title-tools title reset" rsf:id="pagetitlelink"><span class="Mrphs-toolTitleNav__link Mrphs-toolTitleNav__link--reset"></span><span rsf:id="pagetitletext"></span></a>
```
Comment (bjones86) [inline lessonbuilder/tool/src/webapp/templates/ShowPage.html]: Whitespace

File: lessonbuilder/tool/src/webapp/templates/ShowPage.html [hunk 20]
```diff
@@ -220,7 +220,9 @@ <h3 id="toolbar-label" rsf:id="msg=simplepage.toolbar" class="offscreen"></h3>
 	var breadcrumbs = jQuery(".breadcrumbs span");
 	// if we're in morpheus, move breadcrums into top bar, and generate an H2 with the title
 	if (jQuery(".oldPortal").size() === 0) {
-	    var title = jQuery(".title span:last").text();
+        var title = jQuery(".title span:last").text();
```
Comment (bjones86) [inline lessonbuilder/tool/src/webapp/templates/ShowPage.html]: Whitespace

File: lessonbuilder/tool/src/webapp/templates/ShowPage.html [hunk 21]
```diff
@@ -410,6 +412,7 @@ <h3 rsf:id="sectionHeader" id="sectionHeader" class="sectionHeader"><span rsf:id
                         <div rsf:id="link-div:" class="link-div">
                           <span rsf:id="item-icon" aria-hidden="true" class="fa-item-text"></span>
                           <span rsf:id="link-note" class="offscreen"></span><span class="link-div-head" role="heading" aria-level="3"><a class="itemlink" href="#" rsf:id="link"><span rsf:id="link-text" class="link-text itemtext"></span></a></span>
+                            <span rsf:id="owner-text" class=""></span>
```
Comment (bjones86) [inline lessonbuilder/tool/src/webapp/templates/ShowPage.html]: Whitespace

**Comment by nicholaswilson100** [comment]
@bjones86 Thanks loads for so much detail in your comments. I'll make the changes today.

File: lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/producers/ShowPageProducer.java (hunk 16 above)
Comment (nicholaswilson100) [inline lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/producers/ShowPageProducer.java]: I'm not sure what you mean for this one.  Do you want the empty line removed?  Everything else is consistently tab-indented.

**Comment by nicholaswilson100** [comment]
I've made the changes.

**Comment by bjones86** [comment]
Thanks, @nicholaswilson100, for addressing those so quickly

**Comment by nicholaswilson100** [comment]
@clhedrick : What do you think of this? The requirements have changed a bit from Neelam's...

File: lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/producers/ShowPageProducer.java [hunk 22]
```diff
@@ -3569,7 +3586,25 @@ protected static boolean makeLink(UIContainer container, String ID, SimplePageIt
...
+
+		if (owner== null) {
+			return "";
+		}
+
+		User user = null;
+		try {
+			user = UserDirectoryService.getUser(owner);
+		} catch (UserNotDefinedException e) {
+			log.warn("Owner #: " + owner + " does not have an associated user.");
```
Comment (buckett) [inline lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/producers/ShowPageProducer.java]: Shouldn't be a warn as user's getting deleted is expected and the content will still be owned.

File: lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/producers/ShowPageProducer.java [hunk 23]
```diff
@@ -3390,6 +3400,13 @@ protected static boolean makeLink(UIContainer container, String ID, SimplePageIt
 					link.decorate(new UIFreeAttributeDecorator("lessonbuilderitem", itemString));
 					fakeDisableLink(link, messageLocator);
 				}  // otherwise fake
+
+				if (canEditPage || (p.getOwner()!=null && p.getOwner().equals(simplePageBean.getCurrentUserId()))) {
+					String owner = getUserDisplayName(p.getOwner());
+					if (StringUtils.isNotBlank(owner)) {
+						UIOutput.make(container, "owner-text", "(owner: " + owner + ")");
```
Comment (buckett) [inline lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/producers/ShowPageProducer.java]: i18n: owner should come from the resource bundle.

**Comment by jonespm** [comment]
I don't really like the usage naming of the new Boolean column "owned". It seems like you're really just using this to check if the page is a student page or not, because previously if "owner" wasn't null then it wasn't a student page. Maybe calling this column studentPage would make more sense. Or maybe there was a better way of indicating that it was a student page, like a prefix to the name? 

**student::student id**

if (owned != null && owned.startsWith("student::"))

Then we probably don't even need this new column and half of this code

**Comment by nicholaswilson100** [comment]
Hi Matthew, I think you're probably right that removal of the 'owned' column would mean that some of the code could be removed but there'd also have to be some changes made all over the place to accommodate the fact that we now need to check if 'student:' is in the 'owner' column,

**Comment by nicholaswilson100** [comment]
So I'm not sure that this change is really worth it in terms of removing code.  It makes sense to me to  call this column 'studentPage' though - as that is what it more accurately indicates than 'owned'.  That wouldn't entail too much of a change I don't think.

**Comment by buckett** [comment]
Just been chatting with @jonespm and having an improved column name be better would be an acceptable improvement to get this merged.

**Comment by lardo-de-arnaud** [comment]
nick, can you take matt J's suggestion and move this forward. 

We know Chuck H is as keen as mustard with this addition and we also know he gave lots of guidance as to how we should implement it in the early days. We cant wait forever for him to comment on this PR so lets leave it 'til (say) next wednesday (8th Feb) and go ahead and merge if he hasnt got back to us. 

Does that sound OK to everybody?

**Comment by lardo-de-arnaud** [comment]
@clhedrick Are you able to look over this?

**Comment by clhedrick** [comment]
This touches enough things that I can't tell from reading it whether it's going to cause issues. The one thing I notice is that it would cause students uploading content for a student page to upload it into a directory within the site collection. The current code causes them to upload it into their own worksite.

The advantage is that it makes the content of pages stay there even if the student leaves the course. The main disadvantage I can see is that it counts against the site quota rather than the user's quota.
Maybe that's not a problem, but sites should be aware of the change.

**Comment by clhedrick** [comment]
Why did you change getEditPrivs? The new code looks OK, but editprivs are cached and you added a check before checking the cached value. I would have thought isPageOwner(page) would have caught the case you're interested in.

**Comment by clhedrick** [comment]
It may not be worth changing, but in editTitle, you allow anyone who is an owner to change the owner. I probably would prohibit this for a student page. There's no clear security implications. It could just break expectations.

**Comment by nicholaswilson100** [comment]
I've had another think (after trying the change) and I'm not so sure the 'owned' column should be renamed 'isStudentPage' for a couple of reasons:

1.  It's not just a simple renaming column change.   'owned' and 'isStudentPage' are not the same thing.  At the moment, the 'owned' column means 'is owned by a page owner' and is by default false and is changed to true if the page is given an owner.  If we changed it to 'isStudentPage', then the column would (presumably) still be default false, but it would be changed to be true *when a student page was created* and so would have very different logic in the code - there would therefore be quite a few code changes and re-testing because of the inverse logic.  

2.   Moreover, we'd need a migration script to change the existing student pages to true which seems like an unnecessary thing to have to include for this new feature which doesn't involve student pages really. 

3.  Most importantly, changing the column name would also mean the column's logic so it would work differently and not in a good way.  If the column's meaning were changed to 'is a student page' then if another feature came along (another type of lesson's page) which worked a bit like student pages but wasn't student pages (call it 'isstudentpage2'), which also gave a type of owners to pages, then we'd have to create another column 'isStudentPage2' to differentiate it from being 'owned' in my sense.  My column really means 'is owned by a page owner' and not the negative 'is not a student page'.    This is inelegant and more code.  At the moment, we don't need this.    It's better to call it 'owned' or 'isPageOwned'.

**Comment by nicholaswilson100** [comment]
I've made the changes Adam's suggested in the Jira, namely the 'change owner' texts and the permissions text changes.

**Comment by nicholaswilson100** [comment]
I've also rebased the branch so it's mergable.

**Comment by nicholaswilson100** [comment]
I didn't make the Index of pages changes Adam and Chuck were suggesting because both he and Chuck  seem to be saying that it's not a 100% necessary thing to get into this PR and it's more important that it's merged.

**Comment by nicholaswilson100** [comment]
On the naming column issue, what I mean is, at the moment a page is owned (in the sense of this PR) iff the owner is not null and owned is true.  If we changed the column to 'isStudentPage' then a  page is owned (in the sense of this PR) iff the owner is not null and isStudentPage is false - but then also if there's another ferature of owned pages (call it 'isOwnedPageSense2') then a page is owned (in the sense of this PR) iff the owner is not null and isStudentPage is false  and isOwnedPageSense2 is false  - not very encapsulated.  This feature shouldn't really be affected significantly by someone else adding another new feature.

**Comment by buckett** [comment]
@nicholaswilson100 Any comments about the getEditPrivs() change?

**Comment by nicholaswilson100** [comment]
I'll take a look...

**Comment by lardo-de-arnaud** [comment]
is this good to be merged now?

**Comment by nicholaswilson100** [comment]
I think I've addressed all the comments in this PR that need addressing.

**Comment by lardo-de-arnaud** [comment]
this has conflicts

**Comment by jonespm** [comment]
These 3 that change lessons will need to be resolved one at a time and merged individually because they will keep conflicting against each other.

**Comment by nicholaswilson100** [comment]
I've rebased this.

**Pull request merged**